File: k-closest-ele.py
# Time Complexity : O(log n-k)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

from math import floor
from typing import List


class Solution:
    def findClosestElements(self, arr: List[int], k: int, x: int) -> List[int]:

        # setting the pointers
        low = 0
        high = len(arr) - k

        # binary search approach
        while low < high:
            mid = floor((low + high) / 2)

            # finding the distances of the pointers
            dist_low = x - arr[mid]
            dist_high = arr[mid + k] - x

            if dist_low > dist_high:
                low = mid+1
            else:
                high = mid

        result = []

        for i in range(low, low + k):
            result.append(arr[i])

        return result

File: KClosestEle.java
//Time Complexity: O(log(n-k))
//Space Compelxity: O(1)
class Solution {
    public List<Integer> findClosestElements(int[] arr, int k, int x) {
        int n=arr.length;
        int low=0;
        int high=n-k;
        
        while(low<high){
            int mid=low+(high-low)/2;
            int distS=x-arr[mid];
            int distL=arr[mid+k]-x;
            if(distS>distL)
                low=mid+1;
            else
                high=mid;
        }
        List<Integer> result=new ArrayList<>();
        for(int i=low;i<low+k;i++){
            result.add(arr[i]);
        }
        return result;
    }
}
